src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/23.diff.sql

```diff
@@ -0,0 +1,203 @@
+		ON dbo.QuantitySearchParam
+	END;
+
+	IF EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue')
+	BEGIN
+	    EXEC dbo.LogSchemaMigrationProgress 'Dropping IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue'
+		DROP INDEX IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue
+		ON dbo.QuantitySearchParam
+	END;
+
+	-- Update datatype and non-nullable LowValue and HighValue columns
+	EXEC dbo.LogSchemaMigrationProgress 'Updating LowValue as NOT NULL'
+	ALTER TABLE dbo.QuantitySearchParam
+	ALTER COLUMN LowValue decimal(18,6) NOT NULL;
+
+	EXEC dbo.LogSchemaMigrationProgress 'Updating HighValue as NOT NULL'
+	ALTER TABLE dbo.QuantitySearchParam
+	ALTER COLUMN HighValue decimal(18,6) NOT NULL;	
+END;
+GO
+
+-- Recreate dropped indexes
+IF NOT EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'IX_QuantitySearchParam_SearchParamId_QuantityCodeId_LowValue_HighValue')
+BEGIN
+	EXEC dbo.LogSchemaMigrationProgress 'Creating IX_QuantitySearchParam_SearchParamId_QuantityCodeId_LowValue_HighValue'
+
+    CREATE NONCLUSTERED INDEX IX_QuantitySearchParam_SearchParamId_QuantityCodeId_LowValue_HighValue
+	ON dbo.QuantitySearchParam
+	(
+		ResourceTypeId,
+		SearchParamId,
+		QuantityCodeId,
+		LowValue,
+		HighValue,
+		ResourceSurrogateId
+	)
+	INCLUDE
+	(
+		SystemId
+	)
+	WHERE IsHistory = 0
+    WITH (ONLINE=ON) 
+	ON PartitionScheme_ResourceTypeId(ResourceTypeId);
+END;
+
+IF NOT EXISTS (SELECT 1 FROM sys.indexes WHERE name = 'IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Creating IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue'
+	
+    CREATE NONCLUSTERED INDEX IX_QuantitySearchParam_SearchParamId_QuantityCodeId_HighValue_LowValue
+    ON dbo.QuantitySearchParam
+    (
+	    ResourceTypeId,
+	    SearchParamId,
+	    QuantityCodeId,
+	    HighValue,
+	    LowValue,
+	    ResourceSurrogateId
+    )
+    INCLUDE
+    (
+	    SystemId
+    )
+    WHERE IsHistory = 0
+    WITH (ONLINE=ON) 
+    ON PartitionScheme_ResourceTypeId(ResourceTypeId);
+END;
+GO
+
+IF NOT EXISTS (
+    SELECT * 
+	FROM sys.key_constraints 
+	WHERE name='PK_QuantitySearchParam' AND type='PK')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Adding PK_QuantitySearchParam'
+	ALTER TABLE dbo.QuantitySearchParam 
+	ADD CONSTRAINT PK_QuantitySearchParam PRIMARY KEY NONCLUSTERED(ResourceTypeId, SearchParamId, HighValue, LowValue, ResourceSurrogateId)
+	WITH (DATA_COMPRESSION = PAGE, ONLINE=ON) 
+	ON PartitionScheme_ResourceTypeId(ResourceTypeId)
+END;
+GO
+
+/*************************************************************
+ Insert singleValue values into the low and high values for
+ dbo.NumberSearchParam table
+**************************************************************/
+EXEC dbo.LogSchemaMigrationProgress 'Populating LowValue and HighValue in NumberSearchParam if null'
```

Comment:
Just a question for my curiosity, why do we set the values in the table after setting the column the NOT NULL.  Is the result when we set the column not null a default value is populated, which is then used in the index.  After that we set the value and the index is rebuilt?

Reply:
Sure!
So the script is following the below sequence for NumberSearchParam or QuantitySearchParam
1. Backfill lowValue and highValue if its null
2. Drop non-clustered indices which has lowValue and highValue in it since we can't alter columns if they exist
3. Alter columns lowValue and highValue as NOT NULL
4. Create dropped indices and primary keys.